File: src/09.tree/47.173.go
package tree

type BSTIterator struct {
	list  []int
	index int
}

func Constructor(root *TreeNode) BSTIterator {
	b := BSTIterator{
		list:  []int{},
		index: 0,
	}

	b.dfs(root)

	return b
}

func (b *BSTIterator) dfs(node *TreeNode) {
	if node == nil {
		return
	}
	b.dfs(node.Left)
	b.list = append(b.list, node.Val)
	b.dfs(node.Right)
}

func (b *BSTIterator) Next() int {
	val := b.list[b.index]
	b.index++
	return val
}

func (b *BSTIterator) HasNext() bool {
	return b.index < len(b.list)
}

File: src/09.tree/47.173.ts
import { TreeNode } from "./01.100";

export class BSTIterator {
    list: number[];
    index: number;
    constructor(root: TreeNode | null) {
        this.list = [];
        this.index = 0;
        this.dfs(root);
    }

    dfs(node: TreeNode | null) {
        if (!node) {
            return;
        }

        this.dfs(node.left);
        this.list.push(node.val);
        this.dfs(node.right);
    }

    next(): number {
        return this.list[this.index++];
    }

    hasNext(): boolean {
        return this.index < this.list.length;
    }
}
